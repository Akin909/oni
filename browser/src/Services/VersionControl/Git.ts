/**
 * Git.ts
 *
 * Utilities around Git
 */

import { Event, IEvent } from "oni-types"
import * as GitP from "simple-git/promise"

import * as Log from "./../../Log"
import VersionControlProvider, { VCSBranchChangedEvent } from "./VersionControlProvider"

export class GitVersionControlProvider implements VersionControlProvider {
    private _onBranchChange = new Event<VCSBranchChangedEvent>()

    constructor(private _git = GitP) {}

    get onBranchChanged(): IEvent<VCSBranchChangedEvent> {
        return this._onBranchChange
    }

    public async getVCSRoot(): Promise<string | null> {
        try {
            return this._git().revparse(["--show-toplevel"])
        } catch (e) {
            Log.warn(`Git provider unable to find vcs root due to ${e.message}`)
            return null
        }
    }

    public getVCSStatus = async (currentDir: string): Promise<GitP.DiffResult | void> => {
        try {
            const isRepo = await this._git(currentDir).checkIsRepo()
            return isRepo && this._git(currentDir).diffSummary()
        } catch (error) {
            Log.warn(`Git privider unable to get current status because of: ${error.message}`)
        }
    }

    public getLocalVCSBranches = (currentDir?: string): Promise<GitP.BranchSummary> => {
        return this._git(currentDir).branchLocal()
    }

    public getVCSBranch = async (currentDir?: string): Promise<string | void> => {
        try {
            const status = await this._git(currentDir).status()
            return status.current
        } catch (e) {
            Log.warn(`Git Provider was unable to get current status because of: ${e.message}`)
        }
    }

    public async changeVCSBranch(targetBranch: string, currentDir: string): Promise<Error | void> {
        try {
            await this._git(currentDir).checkout(targetBranch)
            this._onBranchChange.dispatch(targetBranch)
        } catch (e) {
            return e
        }
    }
}

export default new GitVersionControlProvider()
