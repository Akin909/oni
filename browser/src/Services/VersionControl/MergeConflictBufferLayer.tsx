import { BufferLayer, BufferLayerRenderContext } from "oni-api"
import * as React from "react"

import styled, { pixel, withProps } from "../../UI/components/common"
import { VersionControlManager } from "./"

interface IProps {
    top: number
    left: number
    height?: number
    bottom?: number
    start?: boolean
    end?: boolean
}

interface IConflictProps {
    sectionStart: number
    sectionEnd: number
    markers: JSX.Element[]
}

const MarkerSection = withProps<IProps>(styled.div).attrs({
    style: (props: IProps) => ({
        top: pixel(props.top),
        left: pixel(props.left),
        height: pixel(props.height),
    }),
})`
    width: 100%;
    border: 1px solid green;
    background-color: green;
    color: white;
    position: absolute;
`

const MarkerBody = withProps<IProps>(styled.div).attrs({
    style: (props: IProps) => ({
        top: pixel(props.top),
        left: pixel(props.left),
        bottom: pixel(props.bottom),
        height: pixel(props.height),
    }),
})`
    width: 100%;
    background-color: rgba(0, 128, 0, 0.21);
    position: absolute;
`

const log = (props: IProps) => {
    // tslint:disable
    if (props.start) {
        console.group("START =================")
        console.log("top: ", props.top)
        console.log("bottom: ", props.bottom)
        console.groupEnd()
    } else if (props.end) {
        console.group("END=================")
        console.log("top: ", props.top)
        console.log("bottom: ", props.bottom)
        console.groupEnd()
    }
    // tslint:enable
}

const ConflictMarker: React.SFC<IProps> = props => {
    log(props)
    const bodyHeight = props.end ? props.bottom - props.top : 0
    return (
        <>
            {props.start && (
                <MarkerSection
                    height={props.height}
                    left={props.left}
                    top={props.top}
                    data-id="conflict-marker"
                >
                    HEAD
                </MarkerSection>
            )}
            {props.end && (
                <>
                    <MarkerBody
                        left={props.left}
                        top={props.top + props.height}
                        height={bodyHeight}
                        data-id="conflict-marker-end"
                    />
                    <MarkerSection left={props.left} height={props.height} top={props.bottom} />
                </>
            )}
        </>
    )
}

export default class MergeConflictBufferLayer implements BufferLayer {
    private readonly markers = {
        start: "<<<<<<<",
        end: "=======",
    }

    constructor(private _vcsManager: VersionControlManager) {
        console.log("this._vcsManager: ", this._vcsManager)
    }

    get id() {
        return "merge-conflicts"
    }

    public render(context: BufferLayerRenderContext) {
        return <>{this._getConflictMarkers(context)}</>
    }

    private _isConflictMarker(line: string) {
        return line.startsWith(this.markers.start)
            ? "start"
            : line.startsWith(this.markers.end)
                ? "end"
                : null
    }

    private _getConflictMarkers(context: BufferLayerRenderContext) {
        const { markers } = context.visibleLines.reduce<IConflictProps>(
            (elements, line, currentLine) => {
                const marker = this._isConflictMarker(line)
                if (marker) {
                    const start = context.bufferToScreen({
                        character: 0,
                        line: currentLine,
                    })
                    if (start) {
                        const position = context.screenToPixel({
                            screenX: start.screenX,
                            screenY: start.screenY,
                        })
                        const top = position.pixelY
                        const left = position.pixelX

                        if (marker) {
                            if (marker === "start") {
                                elements.sectionStart = top
                            } else if (marker === "end") {
                                elements.sectionEnd = top
                            }

                            elements.markers.push(
                                <ConflictMarker
                                    left={left}
                                    end={marker === "end"}
                                    start={marker === "start"}
                                    bottom={elements.sectionEnd}
                                    top={elements.sectionStart}
                                    height={context.fontPixelHeight}
                                />,
                            )
                        }

                        return elements
                    }
                }
                return elements
            },
            { sectionStart: null, sectionEnd: null, markers: [] },
        )
        return markers
    }
}
