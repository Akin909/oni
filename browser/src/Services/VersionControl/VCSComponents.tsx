import * as React from "react"

import styled, { withProps } from "./../../UI/components/common"
import { Icon } from "./../../UI/Icon"

interface ICreateIconArgs {
    type: string
    num: number
}

const BranchContainer = styled.div`
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
`

const BranchText = styled.span`
    min-width: 10px;
    text-align: center;
    padding: 2px 4px 0 0;
`

const BranchNameContainer = styled.span`
    width: 100%;
    margin-left: 4px;
`

const ChangeSpanContainer = withProps<{ type: string }>(styled.span)`
    font-size: "0.7rem";
    padding: "0 0.15rem";
    color: ${({ type }) => (type === "plus" ? "green" : "red")};
`

const ChangeSpan = styled.span`
    padding-left: 0.25rem;
`

export const Branch = (props: { branch: string; children: React.ReactNode }) => (
    <BranchContainer>
        <BranchText>
            <Icon name="code-fork" />
            <BranchNameContainer>
                {`${props.branch} `} {props.children}
            </BranchNameContainer>
        </BranchText>
    </BranchContainer>
)

export const VCSIcon = ({ type, num }: ICreateIconArgs) => (
    <span>
        <ChangeSpanContainer type={type}>
            <Icon name={`${type}-circle`} />
        </ChangeSpanContainer>
        <ChangeSpan>{num}</ChangeSpan>
    </span>
)
