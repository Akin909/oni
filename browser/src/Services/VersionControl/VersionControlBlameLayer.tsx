import { Buffer, BufferLayer } from "oni-api"
import * as React from "react"

import { LayerContextWithCursor } from "../../Editor/NeovimEditor/NeovimBufferLayersView"
import styled, { boxShadow, css, pixel, withProps, darken } from "../../UI/components/common"
import { getTimeSince } from "../../Utility"
import { VersionControlProvider } from "./"
import { Blame } from "./VersionControlProvider"

interface IProps extends LayerContextWithCursor {
    getBlame: (lineOne: number, lineTwo: number) => Promise<Blame>
    timeout: number
    currentLineNumber: number
}
interface IState {
    blame: Blame
    showBlame: boolean
}

interface IContainerProps {
    height: number
    top: number
    left: number
    marginLeft: number
    inline: boolean
}

const inlineStyles = css`
    ${(p: IContainerProps) => `
        height: ${pixel(p.height)};
        line-height: ${pixel(p.height)};
        margin-left: ${pixel(p.marginLeft)};
    `};
`

const hoverStyles = css`
    background-color: ${p => p.theme["editor.hover.contents.background"]};
    padding: 0.5em;
    ${boxShadow};
`

const BlameContainer = withProps<IContainerProps>(styled.div).attrs({
    style: ({ height, top, left }: IContainerProps) => ({
        top: pixel(top),
        left: pixel(left),
    }),
})`
    box-sizing: border-box;
    position: absolute;
    width: auto;
    font-style: italic;
    color: ${p => darken(p.theme["menu.foreground"])};
    ${p => (p.inline ? inlineStyles : hoverStyles)};
`

const BlameDetails = styled.span`
    color: inherit;
    opacity: 0.8;
`

export class VCSBlame extends React.PureComponent<IProps, IState> {
    public state: IState = {
        blame: null,
        showBlame: null,
    }
    private _timeout: any
    private readonly LEFT_OFFSET = 15

    public async componentDidMount() {
        const { cursorLine: line } = this.props
        this.resetTimer()
        await this.updateBlame(line, line + 1)
    }

    public async componentDidUpdate(prevProps: IProps) {
        if (prevProps.cursorLine !== this.props.cursorLine) {
            const { currentLineNumber: line } = this.props
            this.resetTimer()
            await this.updateBlame(line, line + 1)
        }
    }

    public resetTimer = () => {
        clearTimeout(this._timeout)
        this.setState({ showBlame: false })
        this._timeout = setTimeout(() => {
            this.setState({ showBlame: true })
        }, this.props.timeout)
    }

    public calculatePosition() {
        const { cursorLine: bufferLine, currentLineNumber: screenLine } = this.props
        const previousBufferLine = bufferLine - 1
        const currentLine = this.props.visibleLines[screenLine]
        const character = currentLine && currentLine.length
        const canFit = this.canFit()
        const positionToRender = canFit
            ? { line: bufferLine, character }
            : { line: previousBufferLine, character: 0 }
        const position = this.props.bufferToPixel(positionToRender)

        return {
            top: position && canFit ? position.pixelY : !canFit ? position.pixelY - 15 : null,
            left: position ? position.pixelX : null,
        }
    }

    public updateBlame = async (lineOne: number, lineTwo: number) => {
        const outOfBounds = !this.isOutOfBounds(lineOne, lineTwo)
        const blame = !outOfBounds ? await this.props.getBlame(lineOne, lineTwo) : null
        this.setState({ blame })
    }

    public formatCommitDate(timestamp: string) {
        return new Date(parseInt(timestamp, 10) * 1000)
    }

    public isOutOfBounds = (...lines: number[]) => {
        return lines.some(
            line => !line || line > this.props.bottomBufferLine || line < this.props.topBufferLine,
        )
    }

    public getBlameText = () => {
        const { blame } = this.state
        if (!blame) {
            return null
        }
        const { author, hash, committer_time } = blame
        const formattedDate = this.formatCommitDate(committer_time)
        const timeSince = getTimeSince(formattedDate)
        const message = `${author}, ${timeSince} ago, #${hash.slice(0, 4).toUpperCase()}`
        return message
    }

    public canFit = () => {
        const { visibleLines, dimensions, currentLineNumber, fontPixelWidth } = this.props
        const offset = Math.round(this.LEFT_OFFSET / fontPixelWidth)
        const message = this.getBlameText()
        const currentLine = visibleLines[currentLineNumber] || ""
        const canFit = dimensions.width > currentLine.length + message.length + offset
        return canFit
    }

    public componentWillUnmount() {
        clearTimeout(this._timeout)
    }

    public render() {
        const { blame, showBlame } = this.state
        return (
            blame &&
            showBlame && (
                <BlameContainer
                    data-id="vcs.blame"
                    marginLeft={this.LEFT_OFFSET}
                    height={this.props.fontPixelHeight}
                    inline={this.canFit()}
                    {...this.calculatePosition()}
                >
                    <BlameDetails>{this.getBlameText()}</BlameDetails>
                </BlameContainer>
            )
        )
    }
}

export default class VersionControlBlameLayer implements BufferLayer {
    constructor(private _buffer: Buffer, private _vcsProvider: VersionControlProvider) {}

    public getBlame = (lineOne: number, lineTwo: number) => {
        return this._vcsProvider.getBlame({ file: this._buffer.filePath, lineOne, lineTwo })
    }

    get id() {
        return "vcs.blame"
    }

    public render(context: LayerContextWithCursor) {
        const currentLineNumber = context.cursorLine + 1
        const currentLineIndex = currentLineNumber - context.topBufferLine
        const activated = this._vcsProvider && this._vcsProvider.isActivated
        return (
            activated && (
                <VCSBlame
                    {...context}
                    timeout={1000}
                    getBlame={this.getBlame}
                    currentLineNumber={currentLineIndex}
                />
            )
        )
    }
}
