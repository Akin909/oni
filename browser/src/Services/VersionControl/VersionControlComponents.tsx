import * as React from "react"

import styled, { IThemeColors, withProps } from "./../../UI/components/common"
import { Icon } from "./../../UI/Icon"

type ChangeTypes = "change" | "addition" | "deletion"

interface ICreateIconArgs {
    type: ChangeTypes
    num: number
}

const BranchContainer = styled.div`
    height: 100%;
    width: 100%;
    display: flex;
    align-items: center;
`

const BranchText = styled.span`
    min-width: 10px;
    text-align: center;
    padding: 2px 4px 0 0;
`

const BranchNameContainer = styled.span`
    width: 100%;
    margin-left: 4px;
`

const selectColorByType = (type: ChangeTypes, theme: IThemeColors) => {
    switch (type) {
        // case "addition":
        //     return theme["highlight.mode.insert.background"]
        // case "deletion":
        //     return "#ff6961"
        // case "change":
        //     return theme["highlight.mode.operator.foreground"]
        default:
            return ""
    }
}

const ChangeSpanContainer = withProps<{ type: ChangeTypes }>(styled.span)`
    font-size: 0.7rem;
    padding: 0 0.15rem;
    color: ${({ type, theme }) => selectColorByType(type, theme)};
`

const ChangeSpan = styled.span`
    padding-left: 0.25rem;
`

export const Branch = (props: { branch: string; children: React.ReactNode }) => (
    <BranchContainer>
        <BranchText>
            <Icon name="code-fork" />
            <BranchNameContainer>
                {`${props.branch} `} {props.children}
            </BranchNameContainer>
        </BranchText>
    </BranchContainer>
)

const getClassNameForType = (type: ChangeTypes) => {
    switch (type) {
        case "addition":
            return "plus-circle"
        case "deletion":
            return "minus-circle"
        case "change":
        default:
            return "question-circle"
    }
}

export const VCSIcon = ({ type, num }: ICreateIconArgs) => (
    <span>
        <ChangeSpanContainer type={type}>
            <Icon name={getClassNameForType(type)} />
        </ChangeSpanContainer>
        <ChangeSpan>{num}</ChangeSpan>
    </span>
)
