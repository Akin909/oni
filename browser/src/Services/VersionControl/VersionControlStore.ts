import { createStore as createReduxStore } from "./../../Redux"
import { StatusResult } from "./VersionControlProvider"

export interface IState {
    status: StatusResult
    hasFocus: boolean
}

interface IGenericAction<T, P = undefined> {
    type: T
    payload?: P
}

export const DefaultState: IState = {
    status: {
        currentBranch: null,
        staged: [],
        conflicted: [],
        created: [],
        modified: [],
        remoteTrackingBranch: null,
        deleted: [],
        untracked: [],
        ahead: null,
        behind: null,
    },
    hasFocus: null,
}

type IEnterAction = IGenericAction<"ENTER">
type ILeaveAction = IGenericAction<"LEAVE">
type IStatusAction = IGenericAction<"STATUS", { status: StatusResult }>
type IAction = IStatusAction | IEnterAction | ILeaveAction

function reducer(state: IState, action: IAction) {
    switch (action.type) {
        case "ENTER":
            return { ...state, hasFocus: true }
        case "LEAVE":
            return { ...state, hasFocus: false }
        case "STATUS":
            return {
                ...state,
                status: action.payload.status,
            }
        default:
            return state
    }
}

export default createReduxStore("Version Control", reducer, DefaultState)
