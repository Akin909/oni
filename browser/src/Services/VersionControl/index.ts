export {
    default as VersionControlProvider,
    Summary,
    SupportedProviders,
} from "./VersionControlProvider"
export { activate, getInstance, VersionControlManager } from "./VersionControlManager"
export { default as VersionControlPane } from "./VersionControlPane"
export { default as store } from "./VersionControlStore"
export { default as VersionControlView } from "./VersionControlView"
