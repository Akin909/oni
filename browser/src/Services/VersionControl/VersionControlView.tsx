import * as React from "react"
import { connect } from "react-redux"

import { styled, withProps } from "./../../UI/components/common"
import { VimNavigator } from "./../../UI/components/VimNavigator"
import { StatusResult } from "./VersionControlProvider"
import { IState } from "./VersionControlStore"

const Row = styled.div`
    display: flex;
    span >  {
        margin-right: 0.2em;
    }
`

const Column = withProps<{ isSelected: boolean }>(styled.div)`
    display: flex;
    flex-direction: column;
    border: ${p => p.isSelected && `1px solid ${p.theme["highlight.mode.normal.background"]}`};
    padding: 0.3em;
`

const Name = styled.span`
    max-width: 8em;
    white-space: nowrap;
    text-overflow: ellipses;
`

const Title = styled.h4`
    margin: 0;
`

const SectionTitle = styled.div`
    margin: 0.2em 0;
    padding: 0.2em;
    background-color: rgba(0, 0, 0, 0.2);
    display: flex;
    justify-content: space-between;
`

interface IModifiedFilesProps {
    files?: string[]
    title: string
    selectedId: string
    symbol: string
}

const GitStatus = ({ title, files, selectedId, symbol }: IModifiedFilesProps) => (
    <div>
        {files && (
            <div>
                <SectionTitle>
                    <Title>{title}</Title>
                    <strong>{files.length}</strong>
                </SectionTitle>
                {files.map(path => (
                    <Column key={path} isSelected={selectedId === path}>
                        <Name>{path}</Name>
                        <Row>{symbol}</Row>
                    </Column>
                ))}
            </div>
        )}
    </div>
)

interface IProps {
    status: StatusResult
    hasFocus: boolean
    getStatus?: () => void
    handleSelection?: (selection: string) => void
    children?: React.ReactNode
}

interface ILocalState {
    error: boolean
}

class VersionControlContainer extends React.Component<IProps, ILocalState> {
    public state: ILocalState = {
        error: null,
    }

    public async componentDidMount() {
        await this.props.getStatus()
    }

    public async componentDidCatch(e: Error) {
        // tslint:disable-next-line
        console.warn(e)
        this.setState({ error: true })
    }

    public render() {
        const { modified, staged, untracked } = this.props.status
        return this.state.error ? (
            <SectionTitle>
                <Title>Something Went Wrong</Title>
            </SectionTitle>
        ) : (
            <VimNavigator
                ids={[...modified, ...untracked, ...staged]}
                active={this.props.hasFocus}
                onSelected={this.props.handleSelection}
                render={selectedId => (
                    <div>
                        <GitStatus
                            selectedId={selectedId}
                            files={modified}
                            title="Modified Files"
                            symbol="M"
                        />
                        <GitStatus
                            selectedId={selectedId}
                            files={staged}
                            title="Staged Files"
                            symbol="S"
                        />
                        <GitStatus
                            selectedId={selectedId}
                            files={untracked}
                            title="Untracked Files"
                            symbol="?"
                        />
                    </div>
                )}
            />
        )
    }
}

export default connect<IState>(
    (state: IState): IProps => ({
        status: state.status,
        hasFocus: state.hasFocus,
    }),
    null,
)(VersionControlContainer)
