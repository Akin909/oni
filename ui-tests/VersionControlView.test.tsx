import { mount, shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import * as React from "react"
import { Provider } from "react-redux"
import configureStore, { MockStore, MockStoreCreator } from "redux-mock-store"

const mockStore: MockStoreCreator<IState> = configureStore()

import { DefaultState, IState } from "./../browser/src/Services/VersionControl/VersionControlStore"
import VersionControlView, {
    GitStatus,
    SectionTitle,
} from "./../browser/src/Services/VersionControl/VersionControlView"

describe("<VersionControlView />", () => {
    const store = mockStore(DefaultState)
    const container = mount(
        <Provider store={store}>
            <VersionControlView getStatus={() => ({})} />
        </Provider>,
    )
    it("renders without crashing", () => {
        expect(container.length).toBe(1)
    })

    it("should render an untracked, staged and modified section", () => {
        const sections = container.find(GitStatus).length
        expect(sections).toBe(3)
    })

    it("shouldn't show a section if it has no content", () => {
        const wrapper = shallow(
            <GitStatus title="modified files" selectedId="file1" symbol="M" files={null} />,
        )
        expect(wrapper.find(SectionTitle).length).toBe(0)
    })

    it("should match the last recorded snapshot unless a change was made", () => {
        const wrapper = shallow(
            <GitStatus
                title="modified files"
                selectedId="file1"
                symbol="M"
                files={["test1", "test2"]}
            />,
        )
        expect(shallowToJson(wrapper)).toMatchSnapshot()
    })

    it("should render the correct number of modified files from the store in the correct section from of the pane", () => {
        const stateCopy = {
            ...DefaultState,
            status: {
                currentBranch: null,
                staged: [],
                conflicted: [],
                created: [],
                modified: ["test1", "test2"],
                remoteTrackingBranch: null,
                deleted: [],
                untracked: [],
                ahead: null,
                behind: null,
            },
        }
        const storeWithMods = mockStore(stateCopy)
        const containerWithMods = mount(
            <Provider store={storeWithMods}>
                <VersionControlView getStatus={() => ({})} />
            </Provider>,
        )

        const modified = containerWithMods.find("[data-test='Modified Files-2'] > strong")
        expect(modified.text()).toBe("2")
    })
})
