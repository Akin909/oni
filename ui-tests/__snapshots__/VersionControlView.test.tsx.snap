// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionControlView /> should match the last recorded snapshot unless a change was made 1`] = `
<div>
  <div>
    <styled.div
      data-test="modified files-2"
    >
      <styled.h4>
        modified files
      </styled.h4>
      <strong>
        2
      </strong>
    </styled.div>
    <styled.div
      isSelected={false}
      key="test1"
    >
      <styled.span>
        test1
      </styled.span>
      <styled.div>
        <strong>
          M
        </strong>
      </styled.div>
    </styled.div>
    <styled.div
      isSelected={false}
      key="test2"
    >
      <styled.span>
        test2
      </styled.span>
      <styled.div>
        <strong>
          M
        </strong>
      </styled.div>
    </styled.div>
  </div>
</div>
`;
